AttachFromFile: rewrite retry loop

tschoonj opened the pull request:
## Description of the Pull Request (PR):

Recently an issue was discovered when using the Singularity container runtime in Arvados Crunch where certain Slurm jobs would start failing with an error message:

```
2021-11-26T18:57:12.918958066Z ERROR [container bcftools_convert_10] (arvc1-xvhdp-ffcjthyczz8gqol) error log:
2021-11-26T18:57:12.918958066Z 
2021-11-26T18:57:12.918958066Z   2021-11-26T18:56:45.272198441Z crunch-run Not starting a gateway server (GatewayAuthSecret was not provided by dispatcher)
2021-11-26T18:57:12.918958066Z   2021-11-26T18:56:45.272309584Z crunch-run crunch-run 2.3.1 (go1.17.1) started
2021-11-26T18:57:12.918958066Z   2021-11-26T18:56:45.272331939Z crunch-run Executing container 'arvc1-dz642-tlt1uw3hyppd06g' using singularity runtime
2021-11-26T18:57:12.918958066Z   2021-11-26T18:56:45.272354161Z crunch-run Executing on host 'slurm-worker-blue-4'
2021-11-26T18:57:12.918958066Z   2021-11-26T18:56:45.433328942Z crunch-run container token "token-obfuscated" 
2021-11-26T18:57:12.918958066Z   2021-11-26T18:56:45.433798736Z crunch-run Running [arv-mount --foreground --read-write --storage-classes default --crunchstat-interval=10 --file-cache 268435456 --mount-by-pdh by_id --disable-event-listening --mount-by-id by_uuid /tmp/crunch-run.arvc1-dz642-tlt1uw3hyppd06g.1063908694/keep2883686336]
2021-11-26T18:57:12.918958066Z   2021-11-26T18:56:49.241177040Z crunch-run Fetching Docker image from collection '3126bcd60bc91b04916bae8dfadede7d+177'
2021-11-26T18:57:12.918958066Z   2021-11-26T18:56:49.372828520Z crunch-run Using Docker image id "sha256:c74eb4247d8550fa2df0c5275a9dd3b34cb105347cc0fcefdf5a05749faaf0a1" 
2021-11-26T18:57:12.918958066Z   2021-11-26T18:56:49.372958779Z crunch-run Loading Docker image from keep
2021-11-26T18:57:12.918958066Z   2021-11-26T18:56:50.079531206Z crunch-run Starting container
2021-11-26T18:57:12.918958066Z   2021-11-26T18:56:50.080267548Z crunch-run Waiting for container to finish
2021-11-26T18:57:12.918958066Z   2021-11-26T18:56:52.939218932Z stderr FATAL:   container creation failed: mount /proc/self/fd/3->/usr/local/var/singularity/mnt/session/rootfs error: while mounting image /proc/self/fd/3: failed to find loop device: could not attach image file to loop device: failed to set loop flags on loop device: resource temporarily unavailable
2021-11-26T18:57:12.918958066Z   2021-11-26T18:56:53.365279723Z crunch-run Complete
2021-11-26T18:57:13.026434934Z ERROR [container bcftools_convert_10] unable to collect output from d41d8cd98f00b204e9800998ecf8427e+0:

```
with version 3.7.4 and 3.9.1 of Singularity. After reporting this issue on the [Arvados bug tracker](https://dev.arvados.org/issues/18489), its developers suggested that the issue is due to Singularity choosing the first `/dev/loopN` device that meets nearly all (but not all) of the criteria it needs:
* exists or can be created
* not already in use
* can be configured with the desired parameters (offset, size limit, flags, etc)

Once it has chosen a device that meets the first two criteria, it waits 1.5s for it to meet the third criterion, then gives up.

The Arvados developers (thanks @tomclegg) have analyzed the underlying issue, and based on their observations we have come up with a patch that does the following:

1. Move the `F_SETFD` and `CmdSetStatus64` syscalls into the "try each /dev/loopN" loop
2. If `CmdSetStatus64` returns `EAGAIN` or `EBUSY`, set `transientErrorFound` to `true`, and move on to try the next loop device
3. When all loop devices have been tried, return a `TransientAttachError`.
4. `AttachFromFile` was renamed to `attachFromFile`, and a new `AttachFromFile` method was written that loops for up to 5 times over `attachFromFile`, while it continues to return a `TransientAttachError`.

We have not managed to come up with a reproducer for this issue outside of Arvados Crunch, as its occurrence appears to be tied to the fact that the Singularity images are accessed through a Fuse mount (using the `arv-mount` command).


### This fixes or addresses the following GitHub issues:

 No issue was opened.


#### Before submitting a PR, make sure you have done the following:

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR, linted with `make check`  and tested this PR locally with a `make test`, and `make testall` if possible (see CONTRIBUTING.md).
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

CC @jrandall

dtrudg commented:
Hello @tschoonj and thanks for the contribution. Please could you squash this down from 12 commits into 1 and update the PR?

I don't have time to look thoroughly today, but should in the next couple of days. There is a consideration of how these change affect retry behavior when Singularity is configures to share loop devices (for processes using the same image) that I want to check. The `EAGAIN` retry has been observed by people making heavy use of shared loop devices, and I need to verify the changes aren't going to drastically affect the sharing in those circumstances.

Thanks again!

tschoonj commented:
@dtrudg thanks for your message. I have squashed the commits down into a single one.

dtrudg commented on pkg/util/loop/loop.go:
```diff
@@ -77,11 +78,41 @@ type Info64 struct {
+		err := loop.attachFromFile(image, mode, number)
+		if err != nil {
+			_, transient := err.(*TransientAttachError)
+			if !transient {
```
If we hit a transient error here it'd be nice to `Sylog.Debugf` it so we can see EAGAIN/EBUSY occurring if needed.

dtrudg commented on pkg/util/loop/loop.go:
```diff
@@ -77,11 +78,41 @@ type Info64 struct {
+			return nil
+		}
+	}
+	return fmt.Errorf("failed to attach to loop device")
```
If we bubbled up a descriptive wrapped error we can just use it here.

dtrudg reviewed (changes requested):
Thanks for the PR, and many thanks for the detailed review you went through internally, considering the various flows.

I think this is a reasonable fix for the issue you are seeing, and something we could merge.

I have some slight reservations about running through the loop devices rather than retrying on a specific device, as we have had instances where people configure Singularity to use a very restricted max loop devices, rely on shared loop devices etc. It's possible we might come back to it so that in 'shared loop device' mode we have very conservative behaviour... or this simpler / faster behaviour otherwise.

I've left some comments mainly around error handling. I think the error code handling can be simplified and also mad more informative by using the Go 1.13+ error wrapping etc.

tschoonj commented:
@dtrudg many thanks for the thorough review. I hope I addressed your suggestions properly in this latest commit. My go skills are not great, so it is likely I did something wrong.

dtrudg commented:
@tschoonj - the more I look at this code, the more it's something I want to refactor further. Let's merge for the solution to your issue and I'll take that on. Thanks!

tschoonj commented:
Awesome, thanks!
